libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/IDPRequestHandler.java

```diff
@@ -0,0 +1,173 @@
+
+import android.content.Context;
+import android.text.TextUtils;
+import android.webkit.WebView;
+
+import com.salesforce.androidsdk.R;
+import com.salesforce.androidsdk.accounts.UserAccount;
+import com.salesforce.androidsdk.app.SalesforceSDKManager;
+import com.salesforce.androidsdk.auth.OAuth2;
+import com.salesforce.androidsdk.config.BootConfig;
+import com.salesforce.androidsdk.rest.ApiVersionStrings;
+import com.salesforce.androidsdk.rest.ClientManager;
+import com.salesforce.androidsdk.rest.RestClient;
+import com.salesforce.androidsdk.rest.RestRequest;
+import com.salesforce.androidsdk.rest.RestResponse;
+import com.salesforce.androidsdk.util.SalesforceSDKLogger;
+
+import java.io.IOException;
+import java.net.URI;
+
+/**
+ * This class performs requests made by an app acting as the IDP app using the SP app's
+ * config and processes the response received before passing it back to the SP app.
+ *
+ * @author bhariharan
+ */
+public class IDPRequestHandler {
+
+    private static final String TAG = "IDPRequestHandler";
+
+    private SPConfig spConfig;
+    private UserAccount userAccount;
+    private RestClient restClient;
+    private String loginUrl;
+
+    /**
+     * Paramterized constructor.
+     *
+     * @param spConfig SP app's config.
+     * @param userAccount User account.
+     * @throws IDPRequestHandlerException
+     */
+    public IDPRequestHandler(SPConfig spConfig, UserAccount userAccount) throws IDPRequestHandlerException {
+        this.spConfig = spConfig;
+        this.userAccount = userAccount;
+        basicValidation();
+        loginUrl = spConfig.getLoginUrl();
+        if (TextUtils.isEmpty(loginUrl)) {
+            loginUrl = userAccount.getLoginServer();
+        }
+        buildRestClient();
+    }
+
+    /**
+     * Kicks off the 'frontdoor' request in the supplied WebView instance.
+     *
+     * @param accessToken Valid access token.
+     * @param webView WebView instance.
+     * @throws IDPRequestHandlerException
+     */
+    public void makeFrontDoorRequest(String accessToken, WebView webView) throws IDPRequestHandlerException {
+        final Context context = SalesforceSDKManager.getInstance().getAppContext();
+        final URI frontdoorUrl = OAuth2.getIDPFrontdoorUrl(userAccount.getInstanceServer(),
+                accessToken, loginUrl, context.getString(R.string.oauth_display_type),
+                spConfig.getOauthClientId(), spConfig.getOauthCallbackUrl(),
+                spConfig.getOauthScopes(), spConfig.getCodeChallenge());
+        webView.loadUrl(frontdoorUrl.toString());
+    }
+
+    /**
+     * Makes a simple REST call to and returns a valid access token or null if refresh failed.
+     * This method should NOT be called from the main thread since it makes a network request.
+     *
+     * @return Valid access token.
+     * @throws IDPRequestHandlerException
+     */
+    public String getValidAccessToken() throws IDPRequestHandlerException {
```

Comment:
We are doing the same thing in `SalesforceDroidGapActivity.java`, may be should introduce a helper method on RestClient to do that - which we could call ensureValidAccessToken.